sw/device/silicon_creator/mask_rom/sigverify_keys.c

```diff
@@ -158,30 +184,79 @@ static rom_error_t dev_key_is_valid(lifecycle_state_t lc_state,
+    case kLcStateTestUnlocked1:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked1);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateTestUnlocked2:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked2);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateTestUnlocked3:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked3);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateTestUnlocked4:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked4);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateTestUnlocked5:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked5);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateTestUnlocked6:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked6);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateTestUnlocked7:
+      HARDENED_CHECK_EQ(lc_state, kLcStateTestUnlocked7);
+      return key_is_valid_in_lc_state_test(key_type);
+    case kLcStateProd:
+      HARDENED_CHECK_EQ(lc_state, kLcStateProd);
+      return key_is_valid_in_lc_state_prod(key_type, key_index);
+    case kLcStateProdEnd:
+      HARDENED_CHECK_EQ(lc_state, kLcStateProdEnd);
+      return key_is_valid_in_lc_state_prod(key_type, key_index);
+    case kLcStateDev:
+      HARDENED_CHECK_EQ(lc_state, kLcStateDev);
+      return key_is_valid_in_lc_state_dev(key_type, key_index);
+    case kLcStateRma:
+      HARDENED_CHECK_EQ(lc_state, kLcStateRma);
+      return key_is_valid_in_lc_state_rma(key_type, key_index);
     default:
-      return kErrorSigverifyBadKey;
+      HARDENED_UNREACHABLE();
   }
 }
 
 rom_error_t sigverify_rsa_key_get(uint32_t key_id, lifecycle_state_t lc_state,
                                   const sigverify_rsa_key_t **key) {
   const sigverify_mask_rom_key_t *keys = sigverify_rsa_keys_ptr_get();
   size_t num_keys = sigverify_num_rsa_keys_get();
-  for (size_t i = 0; i < num_keys; ++i) {
-    const sigverify_mask_rom_key_t *cand_key = &keys[i];
-    if (sigverify_rsa_key_id_get(&cand_key->key.n) == key_id) {
-      RETURN_IF_ERROR(key_is_valid(cand_key->key_type, lc_state, i));
-      *key = &cand_key->key;
+  size_t cand_key_index = UINT32_MAX;
+  size_t i = 0;
+  for (; launder32(i) < num_keys; ++i) {
+    const sigverify_mask_rom_key_t *k = &keys[i];
+    size_t k_id = sigverify_rsa_key_id_get(&k->key.n);
+    if (launder32(k_id) == key_id) {
+      HARDENED_CHECK_EQ(k_id, key_id);
+      rom_error_t error = key_is_valid(k->key_type, lc_state, i);
+      if (launder32(error) == kErrorOk) {
+        HARDENED_CHECK_EQ(error, kErrorOk);
+        cand_key_index = i;
+      }
+    }
+  }
+  HARDENED_CHECK_EQ(i, num_keys);
+
+  if (launder32(cand_key_index) < num_keys) {
+    HARDENED_CHECK_LT(cand_key_index, num_keys);
+    rom_error_t error =
+        key_is_valid(keys[cand_key_index].key_type, lc_state, cand_key_index);
+    if (launder32(error) == kErrorOk) {
```

Comment:
Should we add...
```c
#define HARDENED_RETURN_IF_ERROR(error_) do {
  rom_error_t error = error_;
  RETURN_IF_ERROR(launder32(error));
  HARDENED_CHECK_EQ(error, kErrorOk);
} while (false);
```
I think this will likely be replaced by other upcoming hardenings... but it still feels good to have. We could similarly have a
```c
#define HARDENED_RETURN_IF_GE32(a_, b_, error_) do {
  if (!(launder32(a_) < b_)) { return error_; }
  HARDENED_CHECK_LT(a_, b_);
} while(false)
```

I'm almost tempted to add a loop version too, e.g. `HARDENED_RANGE(size_t, i, 0, num_keys, {...})` but that may be too cute to be worth it.

Reply:
Thanks for the suggestions @mcy, created #10309.

Your first comment made me realize that we don't actually want the `if` here since we expect the key to be valid in the second check. I'll push an update shortly.